R/glm.R

```diff
@@ -0,0 +1,86 @@
+#' Axing a glm.
+#'
+#' glm objects are created from the base \pkg{stats} package.
+#'
+#' @inheritParams butcher
+#'
+#' @return Axed glm object.
+#'
+#' @examples
+#'
+#' cars_glm <- glm(mpg ~ ., data = mtcars)
+#' cleaned_glm <- butcher(cars_glm, verbose = TRUE)
+#'
+#' @name axe-glm
+NULL
+
+#' Remove the call.
+#'
+#' @rdname axe-glm
+#' @export
+axe_call.glm <- function(x, verbose = FALSE, ...) {
+  old <- x
+  x <- exchange(x, "call", call("dummy_call"))
+
+  add_butcher_attributes(
+    x,
+    old,
+    disabled = c("print()", "summary()"),
+    add_class = FALSE,
+    verbose = verbose
+  )
+}
+
+#' Remove the training data.
+#'
+#' @rdname axe-glm
+#' @export
+axe_data.glm <- function(x, verbose = FALSE, ...) {
+  old <- x
+  x <- exchange(x, "data", data.frame(NA))
+  x <- exchange(x, "y", numeric(0))
+
+  add_butcher_attributes(
+    x,
+    old,
+    verbose = verbose
+  )
+}
+
+#' Remove the environments.
+#'
+#' @rdname axe-glm
+#' @export
+axe_env.glm <- function(x, verbose = FALSE, ...) {
+  old <- x
+  x$terms <- axe_env(x$terms, ...)
+  x$family$variance <- axe_env(x$family$variance, ...)
+  x$family$dev.resids <- axe_env(x$family$dev.resids, ...)
+  x$family$aic <- axe_env(x$family$aic, ...)
+  x$family$validmu <- axe_env(x$family$validmu, ...)
+  attributes(x$model)$terms <- axe_env(attributes(x$model)$terms, ...)
+
+  add_butcher_attributes(
+    x,
+    old,
+    verbose = verbose
+  )
+}
+
+#' Remove fitted values.
+#'
+#' @rdname axe-glm
+#' @export
+axe_fitted.glm <- function(x, verbose = FALSE, ...) {
+  old <- x
+  x <- exchange(x, "fitted.values", numeric(0))
+  x <- exchange(x, "residuals", numeric(0))
+  x$qr <- exchange(x$qr, "qr", matrix(0))
```

Comment:
I _think_ we can remove the `fitted.values`, it disables `predict()` with no `newdata` but we typically don't want people to do that.

I don't think we can remove `residuals` or `qr`, see below:

``` r
library(butcher)

glm_fit <- glm(mpg ~ ., data = mtcars)
x <- butcher(glm_fit)

# this doesnt work, but i think we are ok with that
# $fitted.values
predict(x, type = "response")
#> numeric(0)

# I think this fails because `$residuals` and `$qr` are gone.
# axe_fitted.lm doesnt remove these, so we probably shouldnt remove it here.
# maybe add a test for this?
predict(x, newdata = mtcars, se.fit = TRUE)
#> Error in Qr$qr[p1, p1, drop = FALSE]: subscript out of bounds
```

<sup>Created on 2022-03-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Oh bummer, OK 👍